### PR TITLE
Fix migrations

### LINKED ISSUE #94: Pass a Library for all types of  Questions
#### Proposal
When querying questions, I believe that a library will only be present if the question is of type `programming`. Can it be returned for both types of questions (i.e be an invariant like `content`)?

#### Rationale
The MCQ questions will still need a REPL for students to try out code, so we want the REPL to run code in a specific library. Since both Programming and MCQ questions will have a library, we can define the library in the question instead.

### PR DESCRIPTION
**Pull request by tuesmiddt**
THIS IS A BIG AND UGLY PR TO GET THE DB TO [THIS SCHEMA](https://app.sqldbm.com/MySQL/Share/p8t36fKoygIMm98dwKrgxkGFrngIE8md_DYjF4jNYw0).

Fixes #94, #99

TODO: 
- REFACTOR ASSESSMENTS
- REFACTOR GROUPS CONTEXT METHODS AND TESTS

### CODE REVIEW

File: lib/cadet/assessments/question.ex [hunk 1]
```diff
@@ -1,34 +1,36 @@
...
+    field(:type, QuestionType)
     field(:raw_question, :string, virtual: true)
+    embeds_one(:library, Library)
     belongs_to(:assessment, Assessment)
     timestamps()
   end
 
-  @required_fields ~w(title weight question type)a
+  @required_fields ~w(title question type assessment_id)a
   @optional_fields ~w(display_order raw_question)a
```
Comment (indocomsoft) [inline lib/cadet/assessments/question.ex]: `library` + `raw_library`

File: lib/cadet/assessments/submission.ex [hunk 2]
```diff
@@ -21,15 +17,9 @@ defmodule Cadet.Assessments.Submission do
...
-  @optional_fields ~w(override_xp submitted_at)a
-
   def changeset(submission, params) do
-    params = convert_date(params, :submitted_at)
-
     submission
-    |> cast(params, @required_fields ++ @optional_fields)
-    |> validate_required(@required_fields)
+    |> cast(params, [])
     |> validate_role(:student, :student)
```
Comment (indocomsoft) [inline lib/cadet/assessments/submission.ex]: We're moving role validation up to context right? If so maybe can just remove these

File: lib/cadet/course/group.ex [hunk 3]
```diff
@@ -8,14 +8,14 @@ defmodule Cadet.Course.Group do
   alias Cadet.Accounts.User
```
Comment (indocomsoft) [inline lib/cadet/course/group.ex]: Are we moving this to the `Account` context?

File: lib/cadet/course/group.ex [hunk 4]
```diff
@@ -8,14 +8,14 @@ defmodule Cadet.Course.Group do
...
     belongs_to(:leader, User)
-    belongs_to(:student, User)
+    belongs_to(:mentor, User)
+    has_many(:students, User)
   end
 
   def changeset(group, attrs \\ %{}) do
     group
     |> cast(attrs, [])
-    |> foreign_key_constraint(:leader_id)
```
Comment (indocomsoft) [inline lib/cadet/course/group.ex]: I think we should add `foreign_key_constraint` to all the associations?

File: lib/cadet/assessments/question.ex [hunk 5]
```diff
@@ -1,34 +1,37 @@
...
   schema "questions" do
     field(:title, :string)
     field(:display_order, :integer)
-    field(:weight, :integer)
     field(:question, :map)
-    field(:type, ProblemType)
+    field(:type, QuestionType)
     field(:raw_question, :string, virtual: true)
+    field(:max_xp, :integer)
+    embeds_one(:library, Library)
```
Comment (indocomsoft) [inline lib/cadet/assessments/question.ex]: We should add `:raw_library` for json deserialisation

File: lib/cadet/assessments/question.ex [hunk 6]
```diff
@@ -1,34 +1,37 @@
...
     belongs_to(:assessment, Assessment)
     timestamps()
   end
 
-  @required_fields ~w(title weight question type)a
-  @optional_fields ~w(display_order raw_question)a
+  @required_fields ~w(title question type assessment_id)a
+  @optional_fields ~w(display_order raw_question max_xp)a
```
Comment (indocomsoft) [inline lib/cadet/assessments/question.ex]: `library` and `:raw_library`

File: priv/repo/migrations/20180101181301_create_authorizations.exs [hunk 7]
```diff
@@ -17,11 +17,10 @@ defmodule Cadet.Repo.Migrations.CreateAuthorizations do
 
     create(unique_index(:authorizations, [:provider, :uid]))
     create(index(:authorizations, [:provider, :token]))
+    create(index(:authorizations, [:user_id]))
   end
 
   def down do
-    drop(index(:authorizations, [:provider, :uid]))
-    drop(index(:authorizations, [:provider, :token]))
```
Comment (indocomsoft) [inline priv/repo/migrations/20180101181301_create_authorizations.exs]: I think this line should remain and you should add `drop` for the other index and unique_index

File: priv/repo/migrations/20180119002258_create_assessments.exs [hunk 8]
```diff
@@ -1,38 +1,32 @@
...
       add(:is_published, :boolean, null: false)
-      add(:max_xp, :integer)
       add(:priority, :integer)
       timestamps()
     end
 
-    create(index(:assessments, [:order], using: :hash))
     create(index(:assessments, [:open_at]))
     create(index(:assessments, [:close_at]))
   end
```
Comment (indocomsoft) [inline priv/repo/migrations/20180119002258_create_assessments.exs]: Create index `is_published` and `type`

File: priv/repo/migrations/20180704163210_fix_bugged_migrations.exs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+defmodule Cadet.Repo.Migrations.FixBuggedMigrations do
```
Comment (indocomsoft) [inline priv/repo/migrations/20180704163210_fix_bugged_migrations.exs]: Do you wanna just combine this into the respective `create table` instead?

File: lib/cadet/assessments/submission.ex (hunk 2 above)
Comment (tuesmiddt) [reply lib/cadet/assessments/submission.ex]: these are removed in my branch. I don't want to touch it here.

File: lib/cadet/course/group.ex (hunk 4 above)
Comment (tuesmiddt) [reply lib/cadet/course/group.ex]: Add this when we refactor the groups context code.

File: lib/cadet/assessments/answer.ex [hunk 10]
```diff
@@ -5,27 +5,30 @@ defmodule Cadet.Assessments.Answer do
...
 
   @required_fields ~w(answer)a
-  @optional_fields ~w(xp raw_answer)a
+  @optional_fields ~w(xp raw_answer comment adjustment)a
 
   def changeset(answer, params) do
     answer
     |> cast(params, @required_fields ++ @optional_fields)
     |> validate_required(@required_fields)
     |> validate_number(:xp, greater_than_or_equal_to: 0.0)
```
[the remark this answers is not in the document]
Comment (tuesmiddt) [reply lib/cadet/assessments/answer.ex]: changed

File: lib/cadet/assessments/answer.ex [hunk 11]
```diff
@@ -5,27 +5,30 @@ defmodule Cadet.Assessments.Answer do
...
+    |> validate_number(:adjustment, greater_than_or_equal_to: 0.0)
```
[the remark this answers is not in the document]
Comment (tuesmiddt) [reply lib/cadet/assessments/answer.ex]: changed

File: lib/cadet/assessments/assessment.ex [hunk 12]
```diff
@@ -6,29 +6,28 @@ defmodule Cadet.Assessments.Assessment do
...
+    field(:type, AssessmentType)
     field(:summary_short, :string)
     field(:summary_long, :string)
     field(:open_at, Timex.Ecto.DateTime)
     field(:close_at, Timex.Ecto.DateTime)
-    field(:max_xp, :integer, default: 0)
     field(:cover_picture, Image.Type)
     field(:mission_pdf, Upload.Type)
-    field(:order, :string, default: "")
+    field(:priority, :integer)
```
[the remark this answers is not in the document]
Comment (tuesmiddt) [reply lib/cadet/assessments/assessment.ex]: removed

File: lib/cadet/course/group.ex (hunk 3 above)
Comment (tuesmiddt) [reply lib/cadet/course/group.ex]: Based on the current contents and the Course context, I thought it's better to leave it.

File: priv/repo/migrations/20180119002258_create_assessments.exs [hunk 13]
```diff
@@ -1,38 +1,32 @@
...
       add(:summary_short, :text)
       add(:summary_long, :text)
+      add(:type, :assessment_type, null: false)
       add(:open_at, :timestamp, null: false)
       add(:close_at, :timestamp, null: false)
       add(:cover_picture, :string)
       add(:mission_pdf, :string)
       add(:is_published, :boolean, null: false)
-      add(:max_xp, :integer)
       add(:priority, :integer)
```
Comment (indocomsoft) [inline priv/repo/migrations/20180119002258_create_assessments.exs]: This priority should be removed as well

File: lib/cadet/assessments/question.ex (hunk 5 above)
Comment (tuesmiddt) [reply lib/cadet/assessments/question.ex]: as discussed, no need.

File: lib/cadet/assessments/question.ex (hunk 6 above)
Comment (tuesmiddt) [reply lib/cadet/assessments/question.ex]: these shouldn't be here. `cast/4` doesn't work with embeds.

File: priv/repo/migrations/20180101181301_create_authorizations.exs (hunk 7 above)
Comment (tuesmiddt) [reply priv/repo/migrations/20180101181301_create_authorizations.exs]: indexes drop with the table by default
see: https://www.postgresql.org/docs/8.2/static/sql-droptable.html

File: lib/cadet/assessments/question.ex (hunk 6 above)
Comment (indocomsoft) [reply lib/cadet/assessments/question.ex]: I thought you need to cast `:library` first before you can `cast_embed/3`?

Comment (indocomsoft) [reply lib/cadet/assessments/question.ex]: Cos `cast_embed/3` is getting its data from `%Changeset{data: data}`

Comment (indocomsoft) [reply lib/cadet/assessments/question.ex]: You can try it out in the REPL `iex -S mix`

Comment (indocomsoft) [reply lib/cadet/assessments/question.ex]: Never mind, I stand corrected

File: lib/cadet/course/group.ex (hunk 4 above)
Comment (indocomsoft) [reply lib/cadet/course/group.ex]: Ok, #97

File: priv/repo/migrations/20180101181301_create_authorizations.exs (hunk 7 above)
Comment (indocomsoft) [reply priv/repo/migrations/20180101181301_create_authorizations.exs]: Cool

File: lib/cadet/assessments/assessments.ex [hunk 14]
```diff
@@ -86,6 +86,7 @@ defmodule Cadet.Assessments do
     Repo.get(Assessment, id)
   end
 
+  # TODO: FIX THIS SHIT
```
[the remark this answers is not in the document]
Comment (indocomsoft) [reply lib/cadet/assessments/assessments.ex]: Bump. Should we create an issue to solve this once this is merged or are you gonna solve this in this PR? @tuesmiddt

File: lib/cadet/assessments/question.ex (hunk 6 above)
Comment (tuesmiddt) [reply lib/cadet/assessments/question.ex]: Also no need to add it to the fields. can check with this:

```
params = %{"title": "hello", "assessment_id": 4, "type": :multiple_choice, "question": %{"content": "hello", choices: [%{"content": "world", "choice_id": 0, "is_correct": true}]}, "library": %{"version": 0, "globals": ["hello", "world"]}}

Cadet.Assessments.Question.changeset(%Cadet.Assessments.Question{}, params)
```

File: lib/cadet/assessments/assessments.ex (hunk 14 above)
[the remark this answers is not in the document]
Comment (tuesmiddt) [reply lib/cadet/assessments/assessments.ex]: it's meant to refer to a generic assessments refactor. I guess we already have an issue for that so i'll remove it. no functional problems right now

File: lib/cadet/assessments/question.ex (hunk 6 above)
Comment (indocomsoft) [reply lib/cadet/assessments/question.ex]: Yup, I derped. Sorry :/

**Comment by indocomsoft** [comment]
@ningyuansg @remo5000 @sreycodes Please note that this PR has various changes to past migration -- we intend this to be the last PR that requires you to run `mix ecto.reset` and `MIX_ENV=test mix ecto.reset`.

**Pull request merged**